src/GraphQL/Subscription/ObservableExtensions.cs

```diff
@@ -0,0 +1,253 @@
+            {
+                _token = _cancellationTokenSource.Token;
+                _observer = observer;
+                _disposeAction = disposeAction;
+                // ensure that the transform cannot directly throw an exception without it being wrapped in a Task<Exception>
+                _transformError = async (exception, token) => await transformError(exception, token).ConfigureAwait(false);
+                // ensure that the transform cannot directly throw an exception without it being wrapped in a Task<TOut>
+                _transformNext = async (data, token) => await transformNext(data, token).ConfigureAwait(false);
+            }
+
+            public void OnNext(TIn value) => QueueEvent(QueueType.Data, _transformNext(value, _token), null);
+
+            public void OnError(Exception error) => QueueEvent(QueueType.Error, null, _transformError(error, _token));
+
+            public void OnCompleted() => QueueEvent(QueueType.Completion, null, null);
+
+            /// <summary>
+            /// Queues the specified event and if necessary starts watching for an event to complete.
+            /// </summary>
+            private void QueueEvent(QueueType queueType, Task<TOut>? task, Task<Exception>? error)
+            {
+                var queueData = new QueueData { Type = queueType, Data = task, Error = error };
+                bool attach = false;
+                lock (_queue)
+                {
+                    _queue.Enqueue(queueData);
+                    attach = _queue.Count == 1;
+                }
+
+                // start watching for an event to complete, if this is the first in the queue
+                if (attach)
+                {
+                    // for data events, start sending the data once the transformation task completes
+                    // but for error/completion events, send the event now
+                    if (task != null)
+                    {
+                        // start returning data once the first task has completed (or now if already completed)
+                        if (task.IsCompleted)
+                        {
+                            _ = ReturnDataAsync();
+                        }
+                        else
+                        {
+                            _ = task.ContinueWith(_ => ReturnDataAsync());
+                        }
+                    }
+                    else
+                        // start returning data now
+                        _ = ReturnDataAsync();
+                }
+            }
+
+            /// <summary>
+            /// Returns data from the queue in order (or raises errors or completed notifications);
+            /// executes until the queue is empty.
+            /// </summary>
+            private async Task ReturnDataAsync()
```

Comment:
`ReturnDataAsync` naming confuses me. Let's change to `ProcessAllEventsInQueueAsync`

Reply:
Done